pkg/storegateway/indexheader/index/postings.go

```diff
@@ -87,58 +87,105 @@ func newV1PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset
-				}
+	startLen := d.Len()
+	remainingCount := d.Be32()
+	currentKey := ""
+	valuesForCurrentKey := 0
+
+	var lastValuePositions []lastValuePosition
+
+	for d.Err() == nil && remainingCount > 0 {
+		lastKey := currentKey
+		offsetInTable := startLen - d.Len()
+		keyCount := d.Uvarint()
+
+		// The Postings offset table takes only 2 keys per entry (name and value of label).
+		if keyCount != 2 {
+			return nil, errors.Errorf("unexpected key length for posting table %d", keyCount)
+		}
+
+		// Important: this value is only valid as long as we don't perform any further reads from d.
+		// If we need to retain its value, we must copy it before performing another read.
+		key := d.UvarintBytes()
+
+		if len(t.postings) == 0 || currentKey != string(key) {
+			// New label name.
+			currentKey = string(key)
+			t.postings[currentKey] = &postingValueOffsets{}
+			lastValuePositions = append(lastValuePositions, lastValuePosition{key: currentKey, valueOffsetFromEnd: -1})
+			valuesForCurrentKey = 0
+		}
+
+		if valuesForCurrentKey%postingOffsetsInMemSampling == 0 {
+			value := d.UvarintStr()
+			off := d.Uvarint64()
+			t.postings[currentKey].offsets = append(t.postings[currentKey].offsets, postingOffset{value: value, tableOff: offsetInTable})
+
+			if lastKey != currentKey {
 				t.postings[lastKey].lastValOffset = int64(off - crc32.Size)
 			}
-			t.postings[key] = &postingValueOffsets{}
-			valueCount = 0
+
+			// If the current value is the last one for this key, we don't need to record it again.
+			lastValuePositions[len(lastValuePositions)-1].valueOffsetFromEnd = -1
+		} else {
+			// We only need to store this value if it's the last one for this key.
+			// Record our current position in the table and come back to it below if it turns out this is the last value.
+			lastValuePositions[len(lastValuePositions)-1].valueOffsetFromEnd = d.Len()
+			lastValuePositions[len(lastValuePositions)-1].offsetInTable = offsetInTable
+
+			// Skip over the value and offset.
+			d.SkipUvarintBytes()
+			d.Uvarint64()
 		}
 
-		lastKey = key
-		lastValue = value
-		lastTableOff = tableOff
-		valueCount++
+		valuesForCurrentKey++
+		remainingCount--
+	}
 
-		if (valueCount-1)%postingOffsetsInMemSampling == 0 {
-			t.postings[key].offsets = append(t.postings[key].offsets, postingOffset{value: value, tableOff: tableOff})
+	if d.Err() != nil {
+		return nil, errors.Wrap(d.Err(), "read postings table")
+	}
+
+	d.ResetAt(0)
+
+	// Read through the file once more, reading the values and positions of the last value for each key.
+	for _, pos := range lastValuePositions {
+		if pos.valueOffsetFromEnd == -1 {
+			continue
 		}
 
-		return nil
-	}); err != nil {
-		return nil, errors.Wrap(err, "read postings table")
+		d.Skip(d.Len() - pos.valueOffsetFromEnd)
+		value := d.UvarintStr()
+		t.postings[pos.key].offsets = append(t.postings[pos.key].offsets, postingOffset{value: value, tableOff: pos.offsetInTable})
 	}
+
 	if len(t.postings) > 0 {
```

Comment:
Conceptually, shouldn't go before we do the 2nd scan of the file?

Reply:
I've changed the implementation to run in one pass through the file, so I don't think this is relevant any more.